### PR TITLE
spotify-player: make build options configurable

### PR DESCRIPTION
**Pull request by xyven1**
###### Description of changes

All the cargo build options are surfaced as configurable options in the package.

The package description was also updated as spotify-player is clearly not command driven anymore, it is a TUI. The new description is pulled [directly form the homepage README](https://github.com/aome510/spotify-player#introduction).

Finally, I added myself to maintainers list as to meet requirements for pull request.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/spotify-player/default.nix [hunk 1]
```diff
@@ -7,6 +7,15 @@
...
 , fontconfig
+
+# build options
+, withStreaming ? true
+, withAudioBackend ? "rodio"
+, withMediaControl ? true
+, withLyrics ? false
+, withImage ? false
+, withNotify ? false
+, withSixel ? false
```
Comment (dit7ya) [inline pkgs/applications/audio/spotify-player/default.nix:18]: Why not build with all features by default?

**Comment by xyven1** [comment]
Just bumping this PR as it is still relevant, and becomes more relevant as spotify-player gets updated with more options

File: pkgs/applications/audio/spotify-player/default.nix (hunk 1 above)
Comment (xyven1) [reply pkgs/applications/audio/spotify-player/default.nix:18]: Yeah, I don't know why the maintainer didn't merge the PR. PR now needs to be updated for all the new features of spotify_player, but other than that pretty much nothing has changed. 

Having reasonable defaults, or even setting everything as default does not affect the average consumer at all, it just allows others to override when necessary. Pretty basic feature of flakes, and is the standard for packages with compilation options. Potentially options could be disabled at runtime with config in rust and that is the reason to compile with everything? But still, people then have to deal with home-manager or something like that, and it also means that they have an inflated binary.

**Comment by dit7ya** [comment]
Sorry this PR slipped from me, can you please update it to reflect the latest changes @xyven1 ?

**Comment by xyven1** [comment]
Yep, I'll do that today.

**Comment by xyven1** [comment]
Just noticed something: all of the backends definitely have their own buildInputs (i.e. gstreamer probably needs lib-gstreamer-dev), and that is not currently addressed. If someone knows what each of the dependencies are for each backend, I can throw them in.

**Comment by xyven1** [comment]
@dit7ya Ok, I think its up to date. Only had to edit the compilation option for the daemon. The defaults I provide are now exactly identical to what you have, so there shouldn't be any issue there. Last order of business would be to address [this issue](https://github.com/NixOS/nixpkgs/pull/218394#issuecomment-1679984968).

**Comment by xyven1** [comment]
@dit7ya Just pinging you in case this slipped by you

File: pkgs/applications/audio/spotify-player/default.nix [hunk 2]
```diff
@@ -8,6 +8,16 @@
 , dbus
 , fontconfig
 , libsixel
+
+# build options
+, withStreaming ? true
+, withDaemon ? true
+, withAudioBackend ? "rodio"
```
Comment (dit7ya) [inline pkgs/applications/audio/spotify-player/default.nix:15]: Can we add other possible audio-backends here as a comment?

**Comment by dit7ya** [comment]
> Last order of business would be to address https://github.com/NixOS/nixpkgs/pull/218394#issuecomment-1679984968.

Is this solved?

**Comment by xyven1** [comment]
No, but I think the PR should be merged anyway, maybe disabling the option to select backend for now until someone goes through the effort of figuring out all of the build deps of every backend.

**Comment by xyven1** [comment]
alright @dit7ya it is fully solved now. Build deps are conditionally added, and there is an assert to make sure that people only enter a valid backend.

File: pkgs/applications/audio/spotify-player/default.nix [hunk 3]
```diff
@@ -4,12 +4,30 @@
...
+, withDaemon ? true
+, withAudioBackend ? "rodio" # alsa, pulseaudio, rodio, portaudio, jackaudio, rodiojack, sdl
+, withMediaControl ? true
+, withLyrics ? true
+, withImage ? true
+, withNotify ? true
+, withSixel ? true
 }:
 
+assert lib.assertOneOf "withAudioBackend" withAudioBackend [ "" "alsa" "pulseaudio" "rodio" "portaudio" "jackaudio" "rodiojack" "sdl" "gstreamer" ];
```
Comment (dit7ya) [inline pkgs/applications/audio/spotify-player/default.nix:29]: Should it be built without any audio backends `""`?

File: pkgs/applications/audio/spotify-player/default.nix [hunk 4]
```diff
@@ -30,31 +48,37 @@ rustPlatform.buildRustPackage rec {
...
   buildInputs = [
     openssl
-    alsa-lib
     dbus
     fontconfig
-    libsixel
-  ];
+  ]
+    ++ lib.optionals withSixel [ libsixel ]
+    ++ lib.optionals (withAudioBackend == "alsa") [ alsa-lib ]
```
Comment (dit7ya) [inline pkgs/applications/audio/spotify-player/default.nix:55]: Probably can be merged with the `rodio` option.

File: pkgs/applications/audio/spotify-player/default.nix [hunk 5]
```diff
@@ -30,31 +48,37 @@ rustPlatform.buildRustPackage rec {
...
+    ++ lib.optionals (withAudioBackend != "") [ "${withAudioBackend}-backend" ]
+    ++ lib.optionals withMediaControl [ "media-control" ]
+    ++ lib.optionals withImage [ "image" ]
+    ++ lib.optionals withLyrics [ "lyric-finder" ]
+    ++ lib.optionals withDaemon [ "daemon" ]
+    ++ lib.optionals withNotify [ "notify" ]
+    ++ lib.optionals withStreaming [ "streaming" ]
+    ++ lib.optionals withSixel [ "sixel" ];
 
   meta = with lib; {
```
Comment (dit7ya) [inline pkgs/applications/audio/spotify-player/default.nix:76]: Let's update it to fit the convention of not using `with lib;` here.

**Review by dit7ya — CHANGES_REQUESTED** [review]
Looks good to me, sans the few comments.

**Comment by dit7ya** [comment]
Result of `nixpkgs-review pr 218394` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>spotify-player</li>
  </ul>
</details>

File: pkgs/applications/audio/spotify-player/default.nix (hunk 3 above)
Comment (xyven1) [reply pkgs/applications/audio/spotify-player/default.nix:29]: there is no reason why not. One might want to just use it as a client for spotify connect. It builds and runs afaict

File: pkgs/applications/audio/spotify-player/default.nix (hunk 4 above)
Comment (xyven1) [reply pkgs/applications/audio/spotify-player/default.nix:55]: I think its best to have the audio backends exactly as specified in [librespot documentation](https://github.com/librespot-org/librespot/wiki/Compiling#general-dependencies)

File: pkgs/applications/audio/spotify-player/default.nix (hunk 5 above)
Comment (xyven1) [reply pkgs/applications/audio/spotify-player/default.nix:76]: sure, whats the replacement for that?

Comment (xyven1) [reply pkgs/applications/audio/spotify-player/default.nix]: This should be good now. I can also expand `with lib.maintainers` but that feels unnecessarily verbose.

**Comment by xyven1** [comment]
@dit7ya Should be good to merge now?

**Review by dit7ya — APPROVED** [review]

**Comment by xyven1** [comment]
@drupol Bumping this, as it is good to merge.

**Pull request merged**